lib/connections/limiter.go

```diff
@@ -12,53 +12,197 @@ import (
+	for _, v := range from.Devices {
+		if v.DeviceID != to.MyID {
+			devicesToRemove[v.DeviceID] = true
+		}
+	}
+
+	// Mark devices which should not be removed, create new limiters if needed and assign new limiter rate
+	for _, v := range to.Devices {
+		if v.DeviceID == to.MyID {
+			// This limiter was created for local device. Should skip this device
+			continue
+		}
+
+		readLimiter, okR := lim.getDeviceReadLimiter(v.DeviceID)
+		// Device has not been removed or added, this is no longer a candidate for removal
+		if okR {
+			devicesToRemove[v.DeviceID] = false
+		}
+
+		writeLimiter, okW := lim.getDeviceWriteLimiter(v.DeviceID)
+		if okW {
+			if devicesToRemove[v.DeviceID] {
+				// Device has not been removed or added, we should've
+				// already marked it as false in devicesToRemove
+				panic("broken symmetry in device read/write limiters")
+			}
+		}
+
+		// There was no limiter for this ID in map.
+		// This means that we added this device and we should create new limiter
+		if !okR && !okW {
+			readLimiter = rate.NewLimiter(rate.Inf, limiterBurstSize)
+			writeLimiter = rate.NewLimiter(rate.Inf, limiterBurstSize)
+
+			lim.setDeviceLimiters(v.DeviceID, writeLimiter, readLimiter)
+		} else if !okR || !okW {
+			// One of the read/write limiters is not present while the
+			// corresponding write/read one exists. Something has gone wrong.
+			panic("broken symmetry in device read/write limiters")
+		}
+
+		// limiters for this device are created so we can store previous rates for logging
+		previousReadLimit := readLimiter.Limit()
+		previousWriteLimit := writeLimiter.Limit()
+
+		l.Debugf("okR: %t, okW: %t, prevWriteLim: %d, prevReadLim: %d", okR, okW, previousReadLimit, previousWriteLimit)
+
+		lim.setLimitsForDevice(v, readLimiter, writeLimiter)
+
+		// Nothing about this device has changed. Start processing next device
+		if okR && okW &&
+			readLimiter.Limit() == previousReadLimit &&
+			writeLimiter.Limit() == previousWriteLimit {
+			continue
+		}
+
+		readLimitStr := "is unlimited"
+		if v.MaxRecvKbps > 0 {
+			readLimitStr = fmt.Sprintf("limit is %d KiB/s", v.MaxRecvKbps)
+		}
+		writeLimitStr := "is unlimited"
+		if v.MaxSendKbps > 0 {
+			writeLimitStr = fmt.Sprintf("limit is %d KiB/s", v.MaxSendKbps)
+		}
+
+		l.Infof("Device %s send rate %s, receive rate %s", v.DeviceID, readLimitStr, writeLimitStr)
+	}
+
+	// Delete remote devices which were removed in new configuration
+	for deviceID, shouldBeRemoved := range devicesToRemove {
+		l.Debugf("deviceID: %s, shouldBeRemoved: %t", deviceID, shouldBeRemoved)
+		if shouldBeRemoved {
+			lim.deleteDeviceLimiters(deviceID)
+		}
+	}
+
+	l.Debugln("Rebuild of device limiters map finished")
 }
 
-func (lim *limiter) newWriteLimiter(w io.Writer, isLAN bool) io.Writer {
-	return &limitedWriter{writer: w, limiter: lim, isLAN: isLAN}
+// Compare read/write limits in configurations
+func (lim *limiter) checkDeviceLimits(from, to config.Configuration) bool {
```

Comment:
Again, doesn't need to live on lim.
Also the function name doesn't explain well what this does. Should probably be called deviceLimitsChanged() and invert what it returns.